### PR TITLE
Add support for RPLidarA1

### PR DESCRIPTION
**Pull request by peterbarker**
This PR improves the RPLidar proximity driver in two ways:

- Improves reliability by increasing the serial buffer size (the lidar produce large amount of data which was leading to serial buffer corruption)
- Adds SITL simulator
- Adds support for the S1 model

This has been tested in SITL and on real-hardware including putting barriers in front of the lidar at different positions and confirming MP's proximity viewer correctly displays the expected distances.

### CODE REVIEW

File: libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp [hunk 1]
```diff
@@ -97,271 +103,302 @@ void AP_Proximity_RPLidarA2::update(void)
 // get maximum distance (in meters) of sensor
 float AP_Proximity_RPLidarA2::distance_max() const
 {
-    return 16.0f;  //16m max range RPLIDAR2, if you want to support the 8m version this is the only line to change
+    switch (model) {
+    case Model::UNKNOWN:
+        return 0.0f;
+    case Model::A1:
+        return 8.0f;
+    case Model::A2:
```
Comment (amilcarlucas) [inline libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:111]: There are 8m and 16m A2 models, the comment should be kept

File: libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp [hunk 2]
```diff
@@ -97,271 +103,302 @@ void AP_Proximity_RPLidarA2::update(void)
...
+        return 16.0f;
```
Comment (amilcarlucas) [inline libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:112]:
```suggestion
        return 16.0f; //16m max range RPLIDAR2, if you want to support the 8m version this is the only line to change
```

**Review by amilcarlucas — CHANGES_REQUESTED** [review]

File: libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:112]: Can this be read from the device?

**Comment by rmackay9** [comment]
I've put some fixes to at least get this compiling in the last commit in this branch: https://github.com/rmackay9/rmackay9-ardupilot/commits/peterbarker-rplidar-a1-support

[Wiki setup instructions for the A1 are here](https://ardupilot.org/copter/docs/common-rplidar-a2.html).

BTW, I'm hoping to add support for the RPLidarS1 which has slightly different wiring and runs at 25600 baud
autopilot-rplidars1-wiring

File: libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp (hunk 2 above)
Comment (amilcarlucas) [reply libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:112]: AFAIK, no.

Comment (rmackay9) [reply libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:112]: PRX1_MIN and MAX exist these days though so it should be possible for users to set this manually.  Another alternative is that we create multiple PRX1_TYPE enum values for the various sensor but they all end up using the same driver.

File: libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp (hunk 1 above)
Comment (peterbarker) [reply libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:100]: Are there really 8m ranges out there for the A2?

File: libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp (hunk 2 above)
Comment (amilcarlucas) [reply libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:101]: I sugest keeping the 16m to support most A2 devices, but use

min_range = max(PRX1_MIN, 0.05f)
 max_range = min(PRX1_MAX, 16.0f)

**Comment by rmackay9** [comment]
Additional changes on top of this PR to support the S1 are here https://github.com/rmackay9/rmackay9-ardupilot/commits/peterbarker-rplidar-s1-support2

File: libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp (hunk 2 above)
Comment (peterbarker) [reply libraries/AP_Proximity/AP_Proximity_RPLidarA2.cpp:101]: That's essentially what the current code does.

**Comment by peterbarker** [comment]
I've cherry-picked Randy's changes into here

**Review by rmackay9 — APPROVED** [review]

**Comment by Hwurzburg** [comment]
@rmackay9 that picture shows the wrong lidar...thats an S1 not an S2...I will change for the wiki

**Comment by rmackay9** [comment]
@Hwurzburg, OK thanks.  .. I originaly planned to add support for the S2 but then I was sent an S1 so it is actually the S1 that AP supports not the S2.

**Pull request merged**